### ISSUE TITLE
GenerateProto depends on compile tasks

### ISSUE THREAD

**Issue by rougsig** [google/protobuf-gradle-plugin#578]
**Background**
Test project structure ([link](https://github.com/rougsig/protobuf-gradle-plugin-compile)):
```
├── instances
│   ├── app
├── work-with-proto
│   ├── implementation pure-kotlin
│   │   ├── api lib-core
│   ├── implementation pure-java
│   ├── implementation proto-kotlin
│   │   ├── api lib-core
│   ├── implementation proto-java
│   ├── implementation pure-android
│   ├── implementation proto-android
```

**How it works now**
Protobuf gradle plugin create `protobuf` configuration that extends from `implementation` and `compileOnly`.
It's quite user-friendly, because all you project dependencies provides proto files by default. No need to duplication dependency for `protobuf` configuration. 
i.e
```
dependencies {
  implementation project(":with-proto")
  
  // no need write string below.
  // protobuf configuration will take all implementation dependencies automatically.
  // protobuf project(":with-proto")
}
```

**What wrong**
[Gradle scan](https://scans.gradle.com/s/or2wn7tq5m56w) from test project for generate proto task.
Generate proto task resolves proto configuration. Proto configuration resolves `implementation` and `compileOnly` configuration. As result we have compiled or downloaded all project's dependencies.

This is not good. It costs build time a lot. (that looks real, if project have a lot of dependencies https://github.com/google/protobuf-gradle-plugin/issues/551). 

**Expected result**
Generate proto task does not compile and download all of the project's dependencies. It should only work with dependencies marked as protobuf dependencies.

**Changes**
The `protobuf` configuration does not extends from `implementation` and `compileOnly` configurations.

**Breaking changes**
If `protobuf` configuration does not extends from `implementation` and `compileOnly` configurations, user should declarate all dependencies for generate proto task with `protobuf` configuration. e.i
```
dependencies {
  implementation project(":with-proto")
  
  // User should add all protobuf dependencies with protobuf configuration.
  protobuf project(":with-proto")
}
```

**Comment by ejona86** [comment]
Note that `protobuf` configuration is a completely different beast. `protobuf` configuration is for dependencies for which _you want to generate code_ (i.e., it contains a proto and nobody else is generating code for this proto). `implementation` is how you'd add a dependency for your `.proto` files. In a Java-centric environment, you'd never use `protobuf`.

Note that even if we avoid the unnecessarily dependencies like `:lib-core:compileJava`, those tasks are still a dependency for compileJava and the _generated code_. Consider for example if `:lib-core` has .proto files. Then we need those proto files for GenerateProto and we need the previously-generated code for those proto files for JavaCompile.

**Comment by rougsig** [comment]
Yes, it should work as you describe. The `protobuf` configuration only contains *.proto files, but now `protobuf` contains *.class files and other stuff.

**Comment by ejona86** [comment]
> Yes, it should work as you describe. The protobuf configuration only contains *.proto files, but now protobuf contains *.class files and other stuff.

No, that's not what I'm saying. I'm saying that `protobuf` configuration is like adding to `sourceSet.srcDir`. It is completely unrelated to dependencies in the normal sense.

> now protobuf contains *.class files and other stuff.
 
No matter the approach, .proto dependencies often contain class files, because in the normal case the same .jar files contain the .proto files and .class files.
